### PR TITLE
Decompile most of widget drawing

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
@duncanspumpkin Still need to move functions, but the drawing itself should be complete, except for widget 26 (scrollview). Functions are ordered based on original code, for easier checking.

File: src/openloco/window.cpp [hunk 1]
```diff
@@ -1,37 +1,871 @@
...
+
+    // 0x004CA750
+    void draw_23(gfx::drawpixelinfo_t* dpi, window* window, widget* widget, uint16_t flags)
+    {
+        stringFormatBuffer[0] = '\144';
+        format_string(&stringFormatBuffer[1], widget->text, _commonFormatArgs);
+
+        int16_t cx = widget->left + window->x + 2;
+        int16_t ax = widget->right - widget->left - 4 - 14;
+        cx = cx + (ax / 2);
```
Comment (duncanspumpkin) [inline src/openloco/window.cpp]: Cx assigned but not used

File: src/openloco/graphics/colour.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+using namespace openloco::interop;
+
+namespace openloco::colour
+{
+
+    loco_global<uint8_t[32][8], 0x01136BA0> _colour_map_a;
+    loco_global<uint8_t[32][8], 0x01136C98> _colour_map_b;
+
+    uint8_t get_shade(colour_t colour, uint8_t shade)
```
Comment (duncanspumpkin) [inline src/openloco/graphics/colour.cpp]: Can we not just use the same structures as rct. That's what i was planning on doing as viewport initialise also plays with the colour maps.

Comment (marijnvdwerf) [reply src/openloco/graphics/colour.cpp]: Not a fan of that idea. `ColourMapA` is used so frequently without bounds checking, and the member names don't really make any sense. Plus that loco still has the two arrays seperately.

Comment (duncanspumpkin) [reply src/openloco/graphics/colour.cpp]: In rct2 we had the two structures originally. Then they were combined into one later on.

**Comment by marijnvdwerf** [comment]
Bug remaing is that tabs don't get coloured. Also, scrollview still needs to have scrollbars implemented

**Comment by AaronVanGeffen** [comment]
This is looking very promising already. Nice work! Here's a few quick finds.

- [ ] Tabs aren't coloured properly — you already mentioned this.
- [ ] Tabs strips seem to be drawn one pixel too far down compared to `master`.
- [ ] Scrollable views aren't highlighting selections properly — at least the one in the screenshot below (train cargo) isn't. The tooltip is in the right place; I'm pointing at the last carriage there.

screenshot_20180216_211122

**Comment by marijnvdwerf** [comment]
Tabs aren't draw too low; they are supposed to have the same fill colour as the panel below, and are draw a bit lower to break the top highlight of the panel. Basically like this:
image

**Comment by AaronVanGeffen** [comment]
Ah, I see! Thanks for clarifying.

**Comment by marijnvdwerf** [comment]
For some reason the colour pickers behave weirdly; no idea how I caused that to happen.

image

File: src/openloco/widget.cpp [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+                image -= 1;
+
+                if (!hovered)
+                {
+                    image -= 1;
+                }
+            }
+        }
+
+        if (image & (1 << 30))
```
Comment (duncanspumpkin) [inline src/openloco/widget.cpp]: btr. (1<<30) should be cleared.

File: src/openloco/widget.cpp [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+        if (widget->type != widget_type::wt_7)
+        {
+            widget::sub_4CADE8(dpi, window, widget, colour, enabled, disabled, activated);
+            return;
+        }
+
+        // TODO: Remove image addition
+        uint32_t image = widget->image + 2;
+
+        if (image & (1 << 30))
```
Comment (duncanspumpkin) [inline src/openloco/widget.cpp]: as per below this is inverted. and it should clear (1<<30)

**Comment by duncanspumpkin** [comment]
@marijnvdwerf my comments will fix the colour picker.

**Comment by duncanspumpkin** [comment]
@marijnvdwerf is this ready to merge?

**Comment by marijnvdwerf** [comment]
As far as I'm aware: yes. Haven't done a long playtest myself though, as stability on macOS is not yet that amazing.

**Pull request merged**